pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1208 @@
+
+				Ok((state, None, invest_amount))
+			},
+			InvestState::ActiveSwapIntoPoolCurrency { swap } |
+			InvestState::ActiveSwapIntoForeignCurrency { swap } |
+			// We don't care about `done_amount` until swap into foreign is fulfilled
+			InvestState::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone { swap, .. } => {
+				InvestmentState::<T>::insert(who, investment_id, state.clone());
+				Ok((state, Some(swap), Zero::zero()))
+			},
+			InvestState::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing { swap, invest_amount } |
+			InvestState::ActiveSwapIntoForeignCurrencyAndInvestmentOngoing { swap, invest_amount } |
+			// We don't care about `done_amount` until swap into foreign is fulfilled
+			InvestState::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing { swap,invest_amount, .. } => {
+				InvestmentState::<T>::insert(who, investment_id, state.clone());
+				Ok((state, Some(swap), invest_amount))
+			},
+			InvestState::ActiveSwapIntoPoolCurrencyAndSwapIntoForeignDone { swap, done_amount } => {
+				maybe_executed_decrease = Some((swap.currency_out, done_amount));
+
+				let new_state = InvestState::ActiveSwapIntoPoolCurrency { swap };
+				InvestmentState::<T>::insert(who, investment_id, new_state.clone());
+
+				Ok((new_state, Some(swap), Zero::zero()))
+			},
+			InvestState::ActiveSwapIntoPoolCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing { swap, done_amount, invest_amount } => {
+				maybe_executed_decrease = Some((swap.currency_out, done_amount));
+
+				let new_state = InvestState::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing { swap, invest_amount };
+				InvestmentState::<T>::insert(who, investment_id, new_state.clone());
+
+				Ok((new_state, Some(swap), invest_amount))
+			},
+			InvestState::SwapIntoForeignDone { done_swap } => {
+				maybe_executed_decrease = Some((done_swap.currency_in, done_swap.amount));
+
+				InvestmentState::<T>::remove(who, investment_id);
+
+				Ok((InvestState::NoState, None, Zero::zero()))
+			},
+			InvestState::SwapIntoForeignDoneAndInvestmentOngoing { done_swap, invest_amount } => {
+				maybe_executed_decrease = Some((done_swap.currency_in, done_swap.amount));
+
+				let new_state = InvestState::InvestmentOngoing { invest_amount };
+				InvestmentState::<T>::insert(who, investment_id, new_state.clone());
+
+				Ok((new_state, None, invest_amount))
+			},
+		}
+		.map(|(invest_state, maybe_swap, invest_amount)| {
+			// Must update investment amount before handling swap as in case of decrease, 
+			// updating the swap transfers the currency from the investment account to the 
+			// investor which is required for placing the swap order
+			if T::Investment::investment(who, investment_id)? != invest_amount {
+				T::Investment::update_investment(who, investment_id, invest_amount)?;
+			}
+
+			// No need to handle swap order, if redeem state transition is applied afterwards
+			let final_invest_state = if update_swap_order {
```

Comment:
**Maybe-CR**

I think this is not the final value. 
* `Self::handle_swap_order` is updating the state but never returning it, see other comment.